### PR TITLE
[improve][broker] PIP-307: Add proxy support for Java client

### PR DESCRIPTION
**Pull request by dragosvictor**
[PIP-307](https://github.com/apache/pulsar/blob/8beac8b12ef7c0ef54529fbb7e4e76c54dea6283/pip/pip-307.md)

### Motivation

PRs https://github.com/apache/pulsar/pull/21408 and https://github.com/apache/pulsar/pull/21682 implemented producer/consumer support as per the new load balancer transfer mechanism, described by PIP-307, but both assumed that the clients were directly connected to the brokers. This PR implements the respective functionality when connections are made via the Pulsar proxy.

### Modifications

- Proxying information is determined by the `LookupService.getBroker` implementations but is not forwarded to the outside callers. The PR changes this method's signature to return a new `LookupTopicResult` object, that contains all the pertinent information and allows for ease of extension in the future.
- Cache the proxying information (a boolean flag only) in `ConnectionHandler` and use it when the connection is closed with a broker URL redirection.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

  - Added proxy producer/consumer test, verifying that disconnecting the clients respects the assigned URL logical address as well as the proxy physical address. Verified test stability by running it 100 times in the CI runner.
  - Added a test to verify the clients fall back to direct broker connections when the proxy service drops. Note that the test does not mandate a service URL change in the client. Instead, it resolves the URL to the proxy the first time and to the broker the second time. Verified for stability by running it 100 times.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: (https://github.com/dragosvictor/pulsar/pull/3)

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/HttpLookupService.java [hunk 1]
```diff
@@ -101,7 +100,7 @@ public CompletableFuture<Pair<InetSocketAddress, InetSocketAddress>> getBroker(T
                 }
 
                 InetSocketAddress brokerAddress = InetSocketAddress.createUnresolved(uri.getHost(), uri.getPort());
-                return CompletableFuture.completedFuture(Pair.of(brokerAddress, brokerAddress));
+                return CompletableFuture.completedFuture(new LookupTopicResult(brokerAddress, brokerAddress, false));
```
Comment (heesung-sn) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/HttpLookupService.java:103]: nit: can we comment here that the http lookup is always non-proxy?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java [hunk 2]
```diff
@@ -988,6 +990,15 @@ public CompletableFuture<ClientCnx> getConnectionToServiceUrl() {
         return getConnection(address, address, cnxPool.genRandomKeyToSelectCon());
     }
 
+    public CompletableFuture<ClientCnx> getProxiedConnection(final InetSocketAddress logicalAddress,
```
Comment (heesung-sn) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java:993]: nit: can we rename this to `getProxyConnection`?

File: pulsar-proxy/src/test/java/org/apache/pulsar/proxy/server/ProxyWithExtensibleLoadManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        proxyService.start();
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void proxyCleanup() throws Exception {
+        if (proxyService != null) {
+            proxyService.close();
+        }
+    }
+
```
Comment (heesung-sn) [inline pulsar-proxy/src/test/java/org/apache/pulsar/proxy/server/ProxyWithExtensibleLoadManagerTest.java:137]: Could we add a case to assert that the clients can connect to the brokers directly when we shutdown the proxy in the middle,?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java [hunk 4]
```diff
@@ -94,10 +96,13 @@ protected void grabCnx(Optional<URI> hostURI) {
...
-                InetSocketAddress address = InetSocketAddress.createUnresolved(
-                        hostURI.get().getHost(),
-                        hostURI.get().getPort());
-                cnxFuture = state.client.getConnection(address, address, randomKeyForSelectConnection);
+                URI uri = hostURI.get();
+                InetSocketAddress address = InetSocketAddress.createUnresolved(uri.getHost(), uri.getPort());
+                if (useProxy) {
+                    cnxFuture = state.client.getProxiedConnection(address, randomKeyForSelectConnection);
+                } else {
+                    cnxFuture = state.client.getConnection(address, address, randomKeyForSelectConnection);
```
[the remark this answers is not in the document]
Comment (heesung-sn) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java:104]: Sorry. I think `volatile Boolean useProxy` is better to consider thread safety.

File: pulsar-proxy/src/test/java/org/apache/pulsar/proxy/server/ProxyWithExtensibleLoadManagerTest.java (hunk 3 above)
Comment (heesung-sn) [reply pulsar-proxy/src/test/java/org/apache/pulsar/proxy/server/ProxyWithExtensibleLoadManagerTest.java:137]: Let's say a user is upgrading the cluster to migrate from pulsar-proxy to something else(istio). In this case, the connected clients should get connection errors from the proxy (as the proxy is shut down). Then, the clients should try to reconnect and soon get the broker addresses (assuming proxy hosts are removed under the service url and the brokers are added under it).


To simulate this case, I guess we need the following mocks.

0. mock resolveHost() to return the proxy address when called the first time and return the broker address when called the second time.
1. let the client connect to the proxy (assert useProxy is true)
3. shutdown proxy
4. assert that the client connects to the broker (assert useProxy is false)

Comment (heesung-sn) [reply pulsar-proxy/src/test/java/org/apache/pulsar/proxy/server/ProxyWithExtensibleLoadManagerTest.java:137]: However, for istio -> proxy migration, I don't think this graceful reconnection is possible without broker restarts(we need to forcefully unload all clients to reconnect)

**Review by heesung-sn — APPROVED** [review]

**Review by merlimat — APPROVED** [review]

**Pull request merged**